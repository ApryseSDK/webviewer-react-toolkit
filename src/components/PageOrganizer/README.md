> Must use the **Example With State** if you want reorganization on drag to
> work.

Page organizer allows you to view multiple page files and use drag-and-drop to
re-order them.
