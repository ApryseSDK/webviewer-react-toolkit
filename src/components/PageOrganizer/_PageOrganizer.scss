.ui__pageOrganizer {
  background-color: $background-canvas;
}
