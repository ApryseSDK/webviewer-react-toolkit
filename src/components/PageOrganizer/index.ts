export { PageOrganizer as default, PageOrganizerProps } from './PageOrganizer';
