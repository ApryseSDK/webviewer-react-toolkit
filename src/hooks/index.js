'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var tslib_1 = require('tslib');
tslib_1.__exportStar(require('./useAccessibleFocus'), exports);
tslib_1.__exportStar(require('./useCurrentRef'), exports);
tslib_1.__exportStar(require('./useFile'), exports);
tslib_1.__exportStar(require('./useFileSubscribe'), exports);
tslib_1.__exportStar(require('./useFocus'), exports);
tslib_1.__exportStar(require('./useFocusTrap'), exports);
tslib_1.__exportStar(require('./useKeyForClick'), exports);
tslib_1.__exportStar(require('./useManagedFiles'), exports);
tslib_1.__exportStar(require('./useOnClick'), exports);
tslib_1.__exportStar(require('./useToast'), exports);
tslib_1.__exportStar(require('./useUnmountDelay'), exports);
//# sourceMappingURL=index.js.map
