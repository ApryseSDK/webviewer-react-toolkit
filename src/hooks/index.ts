export * from './useAccessibleFocus';
export * from './useCurrentRef';
export * from './useFile';
export * from './useFileSubscribe';
export * from './useFocus';
export * from './useFocusTrap';
export * from './useKeyForClick';
export * from './useManagedFiles';
export * from './useOnClick';
export * from './useToast';
export * from './useUnmountDelay';
