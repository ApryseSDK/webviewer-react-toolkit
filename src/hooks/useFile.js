'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.useFile = void 0;
var react_1 = require('react');
var useFileSubscribe_1 = require('./useFileSubscribe');
/**
 * This hook converts a file class with async values into a React-friendly hook
 * with async values set to undefined until they are fetched.
 * @param file The file to convert to react observable values.
 */
function useFile(file) {
  var _a = (0, useFileSubscribe_1.useFileSubscribe)(
      file,
      function(f) {
        return f.name;
      },
      'onnamechange',
    ),
    name = _a[0],
    nameErr = _a[1];
  var _b = (0, useFileSubscribe_1.useFileSubscribe)(
      file,
      function(f) {
        return f.thumbnail;
      },
      'onthumbnailchange',
    ),
    thumbnail = _b[0],
    thumbnailErr = _b[1];
  var _c = (0, useFileSubscribe_1.useFileSubscribe)(
      file,
      function(f) {
        return f.fileObj;
      },
      'onfileobjchange',
    ),
    fileObj = _c[0],
    fileObjErr = _c[1];
  var _d = (0, useFileSubscribe_1.useFileSubscribe)(
      file,
      function(f) {
        return f.documentObj;
      },
      'ondocumentobjchange',
    ),
    documentObj = _d[0],
    documentObjErr = _d[1];
  var fileValue = (0, react_1.useMemo)(
    function() {
      return {
        file: file,
        id: file.id,
        originalName: file.originalName,
        extension: file.extension,
        name: name,
        thumbnail: thumbnail,
        fileObj: fileObj,
        documentObj: documentObj,
        errors: {
          name: nameErr,
          thumbnail: thumbnailErr,
          fileObj: fileObjErr,
          documentObj: documentObjErr,
        },
      };
    },
    [
      documentObj,
      documentObjErr,
      file,
      fileObj,
      fileObjErr,
      name,
      nameErr,
      thumbnail,
      thumbnailErr,
    ],
  );
  return fileValue;
}
exports.useFile = useFile;
//# sourceMappingURL=useFile.js.map
