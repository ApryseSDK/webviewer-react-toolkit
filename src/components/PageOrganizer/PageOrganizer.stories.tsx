import { boolean } from '@storybook/addon-knobs';
import React, { useCallback, useState } from 'react';
import { File } from '../../hooks/useFile';
import testPdfThumbnailRotated from '../../storybook-helpers/images/pdf-preview-2.png';
import testPdfThumbnail from '../../storybook-helpers/images/pdf-preview.png';
import PageOrganizer from '../PageOrganizer';
import Thumbnail from '../Thumbnail';
import docs from './README.md';

export default { title: 'PageOrganizer', parameters: { info: docs } };

const file: File = {
  id: 'file_1',
  name: 'file_1',
  extension: 'pdf',
  originalName: 'file_1',
  thumbnail: testPdfThumbnail,
};

const fileRotated: File = {
  id: 'file_2',
  name: 'file_2',
  extension: 'pdf',
  originalName: 'file_2',
  thumbnail: testPdfThumbnailRotated,
};

export const Basic = () => {
  const [files, setFiles] = useState([file, fileRotated]);

  const handleOnMove = useCallback((fromIndex: number, toIndex: number) => {
    setFiles(prev => {
      console.log('hey');
      const clone = prev.slice();
      const item = clone.splice(fromIndex, 1)[0];
      clone.splice(toIndex, 0, item);
      console.log();
      return clone;
    });
  }, []);

  return (
    <PageOrganizer
      files={files}
      onMove={handleOnMove}
      selected={boolean('has selected files', false) ? [file.id, fileRotated.id] : undefined}
      onRenderThumbnail={({ file, isSelected, isDragging }) => (
        <Thumbnail file={file} selected={isSelected} style={isDragging ? { opacity: 0.4 } : undefined} />
      )}
    />
  );
};
