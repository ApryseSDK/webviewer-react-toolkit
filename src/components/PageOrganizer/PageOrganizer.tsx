import classnames from 'classnames';
import React, { FC, ReactNode } from 'react';
import { DndProvider } from 'react-dnd';
import Backend from 'react-dnd-html5-backend';
import { File } from '../../hooks/useFile';
import DraggableWrapper from '../DraggableWrapper';

export interface PageOrganizerProps {
  /**
   * A list of files to render out within the page organizer.
   */
  files: File[];
  /**
   * On render function for generating the thumbnails for the page organizer.
   * If you do not want to build your own, try using the `Thumbnail` component.
   */
  onRenderThumbnail: (params: { file: File; isSelected: boolean; isDragging: boolean; index: number }) => ReactNode;
  /**
   * Array of selected files.
   */
  selected?: File['id'][];
  /**
   * Callback fired when a file is moved within the page organizer.
   */
  onMove?: (fromIndex: number, toIndex: number, file: File) => void;
  /**
   * Classname for outer div.
   */
  className?: string;
}

export const PageOrganizer: FC<PageOrganizerProps> = ({ files, selected, onMove, onRenderThumbnail, className }) => {
  const pageOrganizerClass = classnames('ui__base ui__pageOrganizer', className);

  return (
    <div className={pageOrganizerClass}>
      <DndProvider backend={Backend}>
        {files.map((file, index) => (
          <DraggableWrapper
            key={file.id}
            index={index}
            onMove={(fromIndex, toIndex) => onMove?.(fromIndex, toIndex, file)}
            onRenderChildren={isDragging => {
              const isSelected = selected?.includes(file.id) ?? false;
              return onRenderThumbnail({ file, isSelected, isDragging, index });
            }}
          />
        ))}
      </DndProvider>
    </div>
  );
};
