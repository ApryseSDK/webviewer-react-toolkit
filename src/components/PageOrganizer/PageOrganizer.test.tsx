import { shallow } from 'enzyme';
import React from 'react';
import PageOrganizer from '../PageOrganizer';
import { File } from '../../hooks/useFile';

const file: File = {
  id: 'test',
  name: 'test_name',
  extension: 'pdf',
  originalName: 'original_name',
  thumbnail: undefined,
};

describe('PageOrganizer component', () => {
  it('renders its contents', () => {
    const pageOrganizer = shallow(<PageOrganizer files={[file]} onRenderThumbnail={() => 'Thumbnail'} />);
    expect(pageOrganizer.find('.ui__pageOrganizer').length).toEqual(1);
  });

  it('snapshot renders default pageOrganizer', () => {
    const pageOrganizer = shallow(<PageOrganizer files={[file]} onRenderThumbnail={() => 'Thumbnail'} />);
    expect(pageOrganizer).toMatchSnapshot();
  });
});
