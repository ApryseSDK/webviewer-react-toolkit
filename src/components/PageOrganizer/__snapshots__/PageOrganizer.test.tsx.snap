// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageOrganizer component snapshot renders default pageOrganizer 1`] = `
<div
  className="ui__base ui__pageOrganizer"
>
  <DndProvider
    backend={[Function]}
  >
    <Component
      index={0}
      key="test"
      onMove={[Function]}
      onRenderChildren={[Function]}
    />
  </DndProvider>
</div>
`;
